### PR TITLE
Fetch regions from JavaScript API

### PR DESCRIPTION
**Pull request by trevorgerhardt**
Add the endpoint for fetching regions from a local JavaScript API instead of through the Java server. This is a first set of changes that will enable us to see how it works in staging/production. 

All CRUD interactions with a region are now down in the local API (we need a name to distinguish between the two...). We should be able to remove the Region Controller in the Backend after this PR is merged. 

Also, if you look at the files added under `/pages/api/regions` you'll see that it would be quite easy to replicate those routes for all models. There is nothing specific to regions in there. In fact, Next.js allows catch-all API routes that would enable that. So when adding the next model, we would first make those new routes more generic. Then it would all be client side changes.

Additionally, this moves the related Page Components specific for regions directly into their related `/page` file. The idea is that when a React Components that could be re-used, is useful to encapsulate into its own file, or requires its own separate unit tests should live in the `/components` folder. Pages should use those components and should be tested via integration tests. We're slowly moving towards this split.

This PR also:

- Integrates the Cypress tests directly with Cypress service (public repos have a quite generous free plan). Check out the PR comment below for its usefulness.
- Updates the Cypress tests to be more robust using guidelines from Cypress itself. Notably, navigation was unstable.
- Creates a generic "Authenticated Collection" that mimic's the backend's one for using access group permissions.
- Fixes a bug that would show inputs as valid when they weren't on first load (in `/hooks/use-controlled-input`)
- Fixes a bug that would (infrequently) cause map components in to be rendered before the map on initial load. Noticed in Cypress testing, fixed in `/layouts/map`.
- Removed all usage of `Object.assign` because it's not consistently polyfilled across browsers, including Electron.
- Converts many files to TypeScript. (apologies for not consistently using `git mv`)

### CODE REVIEW

File: lib/db/authenticated-collection.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      _id
+    })
+  }
+
+  remove(_id: string) {
+    return this.collection.deleteOne({
+      accessGroup: this.accessGroup,
+      _id
+    })
+  }
```
Comment (ansoncfit) [inline lib/db/authenticated-collection.ts:81]: Do we want to soft-delete regions instead? (Could be tabled for a future change)

**Review by ansoncfit — COMMENTED** [review]
Sorry for the delayed review -- I added a few questions.

File: cypress.json [hunk 2]
```diff
@@ -14,6 +13,7 @@
     "authZeroDomain": "",
     "region": "scratch",
     "dataPrefix": "CYP_",
-    "resetDataBeforeEachRun": true
-  }
+    "resetDataBeforeEachRun": false
```
[the remark this answers is not in the document]
Comment (trevorgerhardt) [reply cypress.json:16]: This flag is for determining whether to reuse scratch data on each Cypress test run or create it from scratch.

File: cypress.json [hunk 3]
```diff
@@ -5,7 +5,6 @@
   "defaultCommandTimeout": 8000,
   "pageLoadTimeout": 120000,
   "retries": 1,
-  "video": false,
```
[the remark this answers is not in the document]
Comment (trevorgerhardt) [reply cypress.json:8]: Agreed 👍

File: lib/auth0.ts [hunk 4]
```diff
@@ -1,33 +1,86 @@
...
-  })
+export async function getSession(req: IncomingMessage): Promise<IUser> {
+  const session = await auth0.getSession(req)
+  if (!session) {
+    throw new Error('User session does not exist. User must be logged in.')
+  }
+
+  const user = {
+    // This is a namespace for a custom claim. Not a URL: https://auth0.com/docs/tokens/guides/create-namespaced-custom-claims
+    accessGroup: session.user['http://conveyal/accessGroup'],
```
[the remark this answers is not in the document]
Comment (trevorgerhardt) [reply lib/auth0.ts:62]: Maybe? It has no effect other than that it's required by Auth0. We have a custom Auth0 rule that adds it (Auth0 -> Auth Pipeline -> Rules -> Move app metadata). I believe this became necessary when we updated the JWT libraries we were using.

File: lib/db/connect.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  throw new Error(
+    'Please define the MONGODB_URL environment variable inside .env.local'
+  )
+}
+
+if (!dbName) {
+  throw new Error(
+    'Please define the MONGODB_DB environment variable inside .env.local'
+  )
+}
```
[the remark this answers is not in the document]
Comment (trevorgerhardt) [reply lib/db/connect.ts:19]: True. I added the default 'analysis' after writing this.

File: lib/db/authenticated-collection.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+  constructor(collection: Collection, session: IUser) {
+    this.accessGroup = getAccessGroup(session)
+    this.collection = collection
+    this.session = session
+  }
+
+  hasAccess(document: void | any) {
+    return document && document.accessGroup === this.accessGroup
+  }
```
[the remark this answers is not in the document]
Comment (trevorgerhardt) [reply lib/db/authenticated-collection.ts:42]: At one point it was but it doesn't appear to be anymore. Will remove 👌

File: lib/db/authenticated-collection.ts (hunk 1 above)
Comment (trevorgerhardt) [reply lib/db/authenticated-collection.ts:77]: We're not currently doing that with regions, so let's table for now. We can make that change later.

File: lib/auth0.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ansoncfit) [reply lib/auth0.ts:62]: Thanks

**Review by ansoncfit — APPROVED** [review]

**Pull request merged**